sklearn/cluster/optics_.py

```diff
@@ -0,0 +1,769 @@
+            DBSCAN `eps` parameter. Must be set to < `max_bound`. Equivalence
+            with DBSCAN algorithm is achieved if `eps` is < (`max_bound` * 5)
+
+        Returns
+        -------
+        core_sample_indices_: array, shape (n_core_samples,)
+            The indices of the core samples.
+
+        labels_ : array, shape (n_samples,)
+            The estimated labels.
+        """
+        check_is_fitted(self, 'reachability_')
+
+        if eps > self.max_bound:
+            raise ValueError('Specify an epsilon smaller than %s. Got %s.'
+                             % (self.max_bound, eps))
+
+        if eps * 5.0 > (self.max_bound * 1.05):
+            warnings.warn(
+                "Warning, max_bound (%s) is close to eps (%s): "
+                "Output may be unstable." % (self.max_bound, eps),
+                RuntimeWarning, stacklevel=2)
+        # Stability warning is documented in extract_dbscan method...
+
+        return extract_dbscan(self.ordering_, self.core_dists_,
+                              self.reachability_, eps)
+
+
+def extract_dbscan(ordering, core_dists, reachability, eps):
+    """Performs DBSCAN extraction for an arbitrary epsilon (`eps`).
+
+    Extraction runs in linear time. Note that if the `max_bound` OPTICS
+    parameter was set to < inf for extracting reachability and ordering
+    arrays, DBSCAN extractions will be unstable for `eps` values close to
+    `max_bound`. Setting `eps` < (`max_bound` / 5.0) will guarantee
+    extraction parity with DBSCAN.
+
+    Parameters
+    ----------
+    ordering: array, shape(n_samples,)
+        OPTICS ordered point indices (`ordering_`)
+    core_dists: array, shaep(n_samples,)
+        Distances at which points become core (`core_dists_`)
+    reachability: array, shape(n_samples,)
+        Reachability distances calculated by OPTICS (`reachability_`)
+    epsilon_prime: float or int
+        DBSCAN `eps` parameter
+
+    Returns
+    -------
+    core_sample_indices_: array, shape (n_core_samples,)
+        The indices of the core samples.
+
+    labels_ : array, shape (n_samples,)
+        The estimated labels.
+    """
+
+    n_samples = len(core_dists)
+    is_core = np.zeros(n_samples, dtype=bool)
+    labels = np.ones(n_samples, dtype=int) * -1
+    cluster_id = 0
+
+    for entry in ordering:
+        if reachability[entry] > eps:
+            if core_dists[entry] <= eps:
+                cluster_id += 1
+                labels[entry] = cluster_id
+            else:
+                # This is only needed for compatibility for repeated scans.
+                labels[entry] = -1   # Noise points
+                is_core[entry] = 0
+        else:
+            labels[entry] = cluster_id
+            if core_dists[entry] <= eps:
+                is_core[entry] = 1   # True
+            else:
+                is_core[entry] = 0   # False
+    return(np.arange(n_samples)[is_core], labels)
+
+
+def extract_optics(ordering, reachability, maxima_ratio=.75,
```

Comment:
same remark.

Reply:
also I don't know how useful it is without the object so the method should be enough

Reply:
Having the function lets users adjust the parameters on the cluster extraction separate from computing the reachability graph... the extraction on the graph is quick, but building the graph takes a while. Since we don't ever change the state of the estimator after it's instantiated, I wrote this as a function that operates on the fitted estimator that has saved the graph. I certainly can make it a class method...but then if a user generates output using the method, the fit parameters stored in the estimator won't match the parameters that generated the labels that they get from the method. The same is true for the function of course, but I guess that it was at least a bit more explicit what was happening since then function forces the user to pass an attribute of the fitted model to a separate function.

I've switched this to private for now, but what's the proper way to implement this? I could add a method interface where it keeps the state the same but returns new labels when called (i.e., similar to the extract_dbscan() method, with the above caveats)... or if you point me in the direction of what sort of documentation/examples are needed for public functions, I can submit a separate PR after this is merged to expose the function publicly. But currently, there's no other way for a user to extract the graph other than re-fitting the estimator...